src/org/javarosa/xform/parse/XFormParser.java

```diff
@@ -480,7 +489,7 @@ This will lead to inconsistency between prefixes used in the form definition (bi
         logger.info(codeTimer.logLine("Creating FormDef from parsed XML"));
     }
 
-    private final Set<String> validElementNames = Collections.unmodifiableSet(new HashSet<>(Arrays.asList(
+    private final Set<String> validElementNames = unmodifiableSet(new HashSet<>(asList(
```

Comment:
What's the difference between the final here and the final removed above? This jarred me a little after reading the commit message. In general, I would have split the changes so the commit message explained the *why* and not the action taken. I would have limited the scope of the change to the semantically relevant pieces only if possible.

Reply:
Thanks for your suggestions!

Regarding `final` here's my rationale:
- Use `final` on all class fields. Remove it if you really need to reassign and there is no alternative
- Avoid `final` elsewhere

I see value expressing that some member is immutable by hinting it with the `final` keyword but my biggest argument is that it forces you to assign a value on construction, which plays really nice when you want to design your state with value objects. I prefer to work with value objects because it's easier to reason about code with them and they often force you to think about  their validity, emptiness, etc., but that's another discussion.

I wouldn't use `final` elsewhere because they tend to hide bigger problems:
- If you use `final` to protect yourself of a too big/complex method, the real solution is to refactor the method.
- If you use `final` to prevent extending classes to override a method, the design on that class hierarchy is probably flawed.

If your methods are small and your class hierarchies are well designed, adding `final` probably makes code harder to read without adding much value to it, in my opinion.

Reply:
>Avoid final elsewhere

You make good points, @ggalmazor. You will probably agree that in a codebase such as this, with some long and confusing methods, with variables in scope for tens of lines, that it’s nice to know (via `final`) that at least a variable’s value doesn’t change.

Reply:
I think they help to understand current code, but if a variable is bound to some meaningful contiguous set of lines, it’s probably better to extract them to a small method.

If the variable is used in many meaningful sets of lines, then extract many small methods and pass the variable as an arg.

If the variable is used with no apparent structure... Then, yeah, `final` brings some good safety restrictions but I think that code would need rework.